docs/signature-layout.md

```diff
@@ -0,0 +1,35 @@
+# Signature File Layout
+
+A common file layout for storing and serving signatures provides a consistent way to reference image signatures. Signatures on a filesystem or a web server shall use this common layout. Signatures stored in a REST API are not required to use this common layout.
+
+## Specification
+
+This specification relies on [RFC3986](https://tools.ietf.org/html/rfc3986), focusing on defining a [path component](https://tools.ietf.org/html/rfc3986#section-3.3) to compose a concise URI reference to a signature.
+
+**SCHEME[AUTHORITY]/PATH_PREFIX/IMAGE@MANIFEST_DIGEST/signature-INT**
+
+**Definitions**
+
+* **SCHEME**: URI scheme per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.1), e.g. **file://** or **https://**
+* **AUTHORITY**: An optional authority reference per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.2), e.g. **example.com**
+* **PATH_PREFIX**: An arbitrary base path to the image component
+* **IMAGE**: The name of the image per [v2 API](https://docs.docker.com/registry/spec/api/#/overview). This would typically take the form of registry/repository/image but is not required to have exactly three parts. There is no requirement to include a **:PORT** component but it should be included if part of the image reference.
```

Comment:
@mhrivnak I’m not sure what issues you see, you must have much more insight into the usual mirroring workflows; so let me try to describe how I understand the situation, to establish a common ground and perhaps explicitly acknowledge some of the downsides:

Really the external/internal use cases and the transparent copy / copy which changes client-visible image ID are intertwined but somewhat orthogonal, I’m afraid the notes below mix the two up a bit but the point should be clear, and perhaps you have a better way to express the matrix of possibilities

> As a user, I want to carefully restrict what software is available to which parts of my infrastructure. As new content is produced by third parties, I want to mirror it on-site, then walk it through a testing workflow before promoting it to a production-facing repository.

(**ISV→customer**) The clients within your infrastructure would refer to the images using the on-site identities, knowing whether they want to use the testing or production images? Then, whatever software is doing the copy from the source to the destination, must also copy the signatures between the source and destination sigstores, and it doesn’t matter whether the sigstore format includes host names because the software doing the copy by definition knows the destination host name and can place the signatures into the correct location.

> … mirror those repositories out to remote locations within my private infrastructure. This is to achieve redundancy and/or scalability, often across multiple facilities.

The difficulty here is in the indirection necessary to convince various clients which conceptually want the same image, and conceptually refer to it using the same ID/location/hostname (`rhel:7.2.3`) to actually talk to the appropriate mirror, with three possible approaches:

- _Transparent_ indirection (different DNS responses for the same host name). No issues here; do a full file-by-file-copy between mirrors, the clients won’t notice. Sigstores work exactly the same as any other mirrors or proxies.
- _A mirror list_ explicitly implemented in the clients, and clients contacting the mirrors instead of the official/primary host name (but this may be system-wide configuration invisible to human users, who still use the official image IDs). Yes, the clients may need a similar client-side mirror configuration system for the sigstores, and yes, that will be a tiny bit more code when the official/primary host names are included in the paths, but overall trivial I think. We don’t support mirror lists for sigstores right now at all, so if/when we do add that, we can handle the hostname component of the path appropriately.
- _User-visible reference changes_, where the image pulling subsystem is not aware of the mirrors, and users, or users’ programs, need to manually update references (instead of pulling `busybox` they need to pull `{Brno,Paris,Dakkar}.mirror/busybox`). Yes, then the host names in sigstore paths would need to be changed when doing copies between mirrors (and that is always possible, this is really the same as the “ISV→customer" case above)—but by far the biggest difficulty is that with changing the image identity like that, the identities in image signatures no longer match (as they shouldn’t). Then it becomes necessary to maintain a specialized `policy.json` for name overrides so that the signatures are accepted, which is awkward all around; it would be overall much preferable to steer users towards the mirror list approach or transparent indirection instead.

I may very well be missing something of course.